pkg/collector/collector.go

```diff
@@ -13,72 +13,93 @@
 )
 
-const usableNo string = "no"
+const (
+	usableNo string = "no"
+)
+
+var domCapabilitiesFilePath = "/etc/kubernetes/node-feature-discovery/source.d/virsh_domcapabilities.xml"
+
+var (
+	baselineCPU = map[string]string{
+		"AMD":   "Opteron_G1",
+		"Intel": "Penryn",
+	}
+
+	notUsableCPUx86 = map[string]bool{
+		"486":        true,
+		"pentium":    true,
+		"pentium2":   true,
+		"pentium3":   true,
+		"pentiumpro": true,
+		"coreduo":    true,
+		"n270":       true,
+		"core2duo":   true,
+		"Conroe":     true,
+		"athlon":     true,
+		"phenom":     true,
+	}
+)
 
 // CollectData retrieves xml data from file and parse them.
-// Output of this function is slice of usable lowercased cpu models.
-// Only models with tag usable yes will be added to slice.
-// <domainCapabilities>
-//   <cpu>
-//     <mode name='custom' supported='yes'>
-//       <model usable='no'>EPYC-IBPB</model>
-//       <model usable='yes'>Haswell</model>
-//     </mode>
-//   </cpu>
-// </domainCapabilities>
-// Output of this xml will be: ["haswell"]
-func CollectData(hostDomCapabilitiesPath string, cpuModelBlackList map[string]bool) ([]string, error) {
+// Output of this function is slice of usable cpu models and features.
+// Only models with tag usable yes will be used.
+func CollectData() ([]string, map[string]bool, error) {
 	hostDomCapabilities := HostDomCapabilities{}
-	err := getStructureFromFile(hostDomCapabilitiesPath, &hostDomCapabilities)
+	err := file.GetStructureFromFile(domCapabilitiesFilePath, &hostDomCapabilities)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
+
+	basicFeaturesMap := make(map[string]bool)
 	cpus := make([]string, 0)
+	features := make(map[string]bool)
 
 	for _, mode := range hostDomCapabilities.CPU.Mode {
+		if mode.Vendor.Name != "" {
+			var err error
+			basicFeaturesMap, err = getBasicFeatures(mode.Vendor.Name)
+			if err != nil {
+				return nil, nil, err
+			}
+
+		}
 		for _, model := range mode.Model {
-			if model.Usable == usableNo || model.Usable == "" {
+			if _, ok := notUsableCPUx86[model.Name]; ok || model.Usable == usableNo || model.Usable == "" {
 				continue
 			}
-			modelName := strings.ToLower(model.Name)
-			if _, ok := cpuModelBlackList[modelName]; !ok {
-				cpus = append(cpus, model.Name)
-			}
+
+			newFeatures, _ := parseFeatures(basicFeaturesMap, model.Name)
+			util.UnionMap(features, newFeatures)
+
+			cpus = append(cpus, model.Name)
 		}
 	}
-	return cpus, nil
+	return cpus, features, nil
 }
 
-//getStructureFromFile load data from file and unmarshals them into given structure
-//Given structure has to be pointer
-func getStructureFromFile(path string, structure interface{}) error {
-	// Open xmlFile
-	fileReader, err := os.Open(path)
+func getBasicFeatures(vendor string) (map[string]bool, error) {
+	fm, err := feature.LoadFeatures(baselineCPU[vendor])
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer fileReader.Close()
+	return fm, nil
+}
```

Comment:
you can probably just call `feature.LoadFeatures`: if err != nil, you should not care about the other return value anyway.

Reply:
I am not sure i fully understand this comment

Reply:
I mean that you can get rid of getBasicFeatures() function entirely. What value is this function adding?